File: internal/compiler/parser.go
package compiler

func Parse(tokens []Token) ASTNode {
	root := ASTNode{
		nodeType: "root",
		value:    "",
		children: []ASTNode{},
	}
	current := root

	for _, token := range tokens {
		switch token.tokenType {
		case "heading":
			root.children = append(root.children, ASTNode{
				nodeType: "heading",
				value:    token.text,
				level:    token.level,
			})
		case "bold":
			root.children = append(root.children, ASTNode{
				nodeType: "bold",
				value:    token.text,
			})
		case "list_item":
			if current.nodeType != "list" {
				listNode := ASTNode{nodeType: "list", children: []ASTNode{}}
				root.children = append(root.children, listNode)
				current = listNode
			}
			current.children = append(current.children, ASTNode{
				nodeType: "list_item",
				value:    token.text,
			})
		default:
			root.children = append(root.children, ASTNode{nodeType: "paragraph", value: token.text})
		}
	}

	return root
}

File: static/script/compiler/html/parser.js
class Parser {
  constructor(tokens) {
    this.tokens = tokens;
    this.position = 0;
  }

  parse() {
    let children = [];
    while (this.position < this.tokens.length) {
      children.push(this.parseNextToken());
    }
    return new DocumentNode(children);
  }

  parseNextToken() {
    const token = this.tokens[this.position++];
    return this.parseToken(token);
  }

  parseToken(token) {
    switch (token.type) {
      case 'HEADING': return new HeadingNode(token.level, token.text);
      case 'BOLD': return new BoldNode(token.text);
      case 'ITALIC': return new ItalicNode(token.text);
      case 'BLOCKQUOTE': return new BlockquoteNode(token.text);
      case 'LISTITEM': return new ListItemNode(token.text);
      case 'CODEBLOCK': return new CodeBlockNode(token.language, token.content);
      case 'INLINECODE': return new InlineCodeNode(token.text);
      case 'TEXT': return new TextNode(token.text, token.inParagraph);
      case 'PARAGRAPH_START': return new ParagraphStartNode();
      case 'PARAGRAPH_CONTENT':
        const childNodes = [];
        token.children.forEach(child => {
          childNodes.push(this.parseToken(child));
        })
        return new ParagraphContentNode(childNodes);
    }
  }
}
